fabric-key-binding-api-v1/src/main/java/net/fabricmc/fabric/api/client/keybinding/v1/FabricKeyBinding.java

```diff
@@ -0,0 +1,204 @@
+ * 			.id(new Identifier("example", "right"))
+ * 			.code(Keys.Right)
+ * 			.build();
+ * KeyBindingRegistry.register(left);
+ * KeyBindingRegistry.register(right);
+ * </code></pre>
+ */
+public class FabricKeyBinding extends KeyBinding {
+	private final Identifier id;
+
+	protected FabricKeyBinding(Identifier id, String translationKey, InputUtil.Type type, int code, String category) {
+		super(translationKey, type, code, category);
+		this.id = id;
+	}
+
+	/**
+	 * Original identifier used to register this key.
+	 *
+	 * <p>May be different from the {@link #getId()}.</p>
+	 */
+	public Identifier getIdentifier() {
+		return id;
+	}
+
+	/**
+	 * Creates a new builder for constructing custom key bindings.
+	 */
+	public static Builder builder() {
+		return new Builder();
+	}
+
+	/**
+	 * Returns the configured KeyCode assigned to the KeyBinding from the player's settings.
+	 *
+	 * @return configured KeyCode
+	 */
+	public InputUtil.KeyCode getBoundKey() {
+		return getBoundKeyOf(this);
+	}
+
+	/**
+	 * Returns the configured KeyCode assigned to the KeyBinding from the player's settings.
+	 *
+	 * @param keyBinding the keybinding
+	 * @return configured KeyCode
+	 */
+	public static InputUtil.KeyCode getBoundKeyOf(KeyBinding keyBinding) {
+		return ((KeyCodeAccessor) keyBinding).getKeyCode();
+	}
+
+	public static class Builder {
+		private static final int UNASSIGNED = InputUtil.UNKNOWN_KEYCODE.getKeyCode();
+
+		private InputUtil.Type type = InputUtil.Type.KEYSYM;
+		private Identifier id = null;
+		private String translationKey;
+		private boolean unassigned = false;
+		private BooleanSupplier toggleFlagSupplier = null;
+		private int code = UNASSIGNED;
+		private String category = KeyCategories.MISC;
+
+		private Builder() {
+		}
+
+		/**
+		 * Sets the key binding id and translation key for this builder.
+		 * <br>
+		 * Key bindings will be assigned a translation key of the format "key.{namespace}.{path}"
+		 *
+		 * @param id Unique identifier for the bound key.
+		 */
+		public Builder id(Identifier id) {
+			this.id = Objects.requireNonNull(id, "Keybinding's id can not be null!");
+			this.translationKey = String.format("key.%s.%s", id.getNamespace(), id.getPath());
+			return this;
+		}
+
+		/**
+		 * Sets the translation key for the key's category. {@link KeyCategories} for
+		 * all possible values, vanilla values.
+		 *
+		 * @param category The category under which key bindings created by this builder will be grouped.
+		 */
+		public Builder category(String category) {
+			this.category = Objects.requireNonNull(category, "Keybinding's category can not be null!");
+			return this;
+		}
+
+		/**
+		 * Sets the default key to be used for the key binding created using this builder.
+		 *
+		 * @param keyCode The default key code. Must be a valid key. May not be -1.
+		 */
+		public Builder code(int keyCode) {
+			Preconditions.checkState(keyCode != UNASSIGNED, "UNASSIGNED is not a valid key code.");
```

Comment:
Is the any way we can verify the integer more? Maybe check against some reasonable lower and upper bound?

Reply:
Just checked, it seems like there isn't a limit on what the integer can be, Minecraft will just create a new code if it doesn't exist already, meaning that it can basically accept any code.